soroban-env-host/src/auth.rs

```diff
@@ -1,3 +1,197 @@
+//! # Auth conceptual overview:
+//!
+//! This module is responsible for two separate tasks (both starting with
+//! "auth"):
+//!
+//!   - Authorization: deciding if some action should be allowed by some policy.
+//!   - Authentication: deciding if some credential or signature is authentic.
+//!
+//! As one would expect, authorization can (though doesn't always) depend on
+//! authentication: part of judging whether some action is allowed may depend on
+//! someone presenting a signed credential, at which point one must evaluate the
+//! credential's authenticity.
+//!
+//! Moreover this subsystem is responsible (as will be discussed in detail
+//! below) with facilitating two different _directions_ for each of these tasks:
+//!
+//!   - Contracts can _require_ auth services provided by this module.
+//!   - Contracts can _provide_ auth services required by this module.
+//!
+//! And again, in both directions the "auth services" required or provided may
+//! be _either_ of authorization, authentication, or both.
+//!
+//! ## Authorization
+//!
+//! Authorization judgments are explicitly requested by smart contracts, and
+//! relate to an ambient authorization context that is maintained by this
+//! module. Specifically a contract calls the host (using `require_auth` or
+//! `require_auth_for_args`) informing the host that the contract code needs
+//! authorization before it continues. If the requirement matches the context
+//! the call succeeds, otherwise it fails (and the contract typically traps).
+//!
+//! This interface raises obvious questions:
+//!
+//!   - What type of "requirements" can the contract require?
+//!   - What are the contents of the authentication context?
+//!
+//! To answer either of these questions we first need to take a slight detour
+//! into the notion of an _address_.
+//!
+//! ## Addresses
+//!
+//! Addresses generalize and abstract over accounts (corresponding to public
```

Comment:
This is not technically incorrect, but I think is a bit misleading. The abstractions here are:

- Accounts (as in 'account abstraction') - entities owned by the users external to the chain. Includes, but not limited to, the Stellar accounts and 'custom' accounts backed by contracts.
- Contracts - entities that only act on-chain (e.g. a liquidity pool contract that transfers money to the user).

The key thing here is that the only special thing about the 'classic' Stellar accounts is that they get a special variant in the `Address` enum, which is more of an implementation detail than anything else.

I think this is correctly reflected in the `Authentication` section below, but here I feel like there is some room for incorrect interpretation.

Reply:
Hm. I think I see what you're saying: that the two different variants of the `Address` type -- `Account` and `Contract` -- do not correspond to _conceptual_ differences between "abstract accounts" and "abstract contracts", because one kind of "abstract account" is a custom account backed by a contract (and so identified by `Address::Contract`). Is that what you mean? I will try to clarify this.

Reply:
Yeah, exactly. The variants basically only influence the authentication code routing and shouldn't otherwise influence any logic.